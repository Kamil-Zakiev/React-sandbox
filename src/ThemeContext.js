import React from 'react'

const Themes = {
    dark: 'dark',
    ligth: 'light',
    oppositeOf: (theme) => {
        return theme === Themes.dark ? Themes.ligth : Themes.dark;
    },
    classFor: (theme) => {
        return theme === Themes.dark ? 'black' : 'lightblue';
    }
};
const ThemeContext = React.createContext(Themes.dark);

export {Themes, ThemeContext};
