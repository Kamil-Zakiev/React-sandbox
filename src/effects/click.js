const CLICK_ANIMATION_FREQ_MS = 15;
const CLICK_DEPTH_PX = 3;
const CLICK_ANIMATION_LENGTH = CLICK_DEPTH_PX * 2;

export default function click({style}) {
    let pos = 0;
    let count = 0;
    let id = setInterval(() => {
        if (count++ === CLICK_ANIMATION_LENGTH) {
            clearInterval(id);
        } else {
            pos += Math.sign(CLICK_DEPTH_PX - count + 0.1); // math magic
            style.top = style.left = pos + "px";
        }
    }, CLICK_ANIMATION_FREQ_MS);
}