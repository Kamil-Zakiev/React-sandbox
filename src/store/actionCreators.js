const ClickCell = (i) => {
    return {
        type: 'CLICK_CELL',
        cell: i
    };
};

const ChangeTheme = () => {
    return {
        type: 'CHANGE_THEME'
    };
};

const ChangeLogDirection = () => {
    return {
        type: 'CHANGE_LOG_DIR'
    };
};

const MoveToStep = (step) => {
    return {
        type: 'MOVE_TO_STEP',
        step: step
    };
};

const SetAllowModal = (isAllowed) => {
    return {
        type: 'SET_ALLOW_MODAL',
        isAllowed: isAllowed
    };
};

export { ClickCell, ChangeTheme, ChangeLogDirection, MoveToStep, SetAllowModal };