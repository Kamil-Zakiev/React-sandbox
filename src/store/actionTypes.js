const CLICK_CELL = 'CLICK_CELL';
const CELL_PRESSED = 'CELL_PRESSED';
const CHANGE_THEME = 'CHANGE_THEME';
const CHANGE_LOG_DIR = 'CHANGE_LOG_DIR';
const MOVE_TO_STEP = 'MOVE_TO_STEP';
const SET_ALLOW_MODAL = 'SET_ALLOW_MODAL';

export {
    CLICK_CELL, 
    CELL_PRESSED,
    CHANGE_THEME, 
    CHANGE_LOG_DIR, 
    MOVE_TO_STEP, 
    SET_ALLOW_MODAL
};
