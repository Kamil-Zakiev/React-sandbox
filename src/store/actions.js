const CLICK_CELL = 'CLICK_CELL';
const CHANGE_THEME = 'CHANGE_THEME';
const CHANGE_LOG_DIR = 'CHANGE_LOG_DIR';
const MOVE_TO_STEP = 'MOVE_TO_STEP';
const SET_ALLOW_MODAL = 'SET_ALLOW_MODAL';

export {
    CLICK_CELL, 
    CHANGE_THEME, 
    CHANGE_LOG_DIR, 
    MOVE_TO_STEP, 
    SET_ALLOW_MODAL
};
