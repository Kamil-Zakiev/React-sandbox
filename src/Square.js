import React, { useRef, useContext, useCallback } from 'react'
import { connect } from 'react-redux'
import { ThemeContext } from './ThemeContext'
import { currentStatus } from './calculateWinner'
import { ClickCell } from './store/actionCreators'
import click from './effects/click';

export function Square({ index, dispatch, inWinDirection, value }) {
    const ref = useRef(null);
    const onClick = useCallback(() => {
        dispatch(ClickCell(index));
        click(ref.current);
    }, [index, dispatch]); // suppose ref.current is always actual so not incuding to dependency list

    const theme = useContext(ThemeContext);
    const className = `square ${inWinDirection && 'winCell'} ${theme}-button`;
    return (
        <button
            ref={ref}
            className={className}
            onClick={onClick}
        >
            {value}
        </button>
    );
}

function mapStateToProps(state, props) {
    const [, , winDirection] = currentStatus(state.history[state.stepNumber]);
    return {
        value: state.history[state.stepNumber].squares[props.index],
        inWinDirection: winDirection && winDirection.indexOf(props.index) !== -1
    }
}

export default connect(mapStateToProps)(Square);