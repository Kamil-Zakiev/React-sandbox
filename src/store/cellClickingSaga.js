import { call, put, takeEvery, race, take, delay, select } from 'redux-saga/effects'
import { CLICK_CELL } from './actions';

function* clickCell(action) {
    yield call(apiSendRequest, action);
    yield put({ type: "CELL_PRESSED" });
}

export function* opponentHasPressed() {
    let opponentClick = yield take(CLICK_CELL);
    while (true) {
        const { clicked } = yield race({
            clicked: take(CLICK_CELL),
            timeout: delay(5000)
        });

        if (!clicked) {
            const opponent = yield select(s => s.history[s.stepNumber].squares[opponentClick.cell]);
            console.log('timeout loose, ', opponent, ' has won!');
            break;
        } else {
            opponentClick = clicked;
        }
    }
}

export function* cellClickingSaga() {
    yield takeEvery(CLICK_CELL, clickCell);
}

function apiSendRequest(action) {
    return new Promise((resolve, reject) => {
        setTimeout(() => {
            console.log('clicked, ', action);
            resolve();
        }, 500);
    });
}