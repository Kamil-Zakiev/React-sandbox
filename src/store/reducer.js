import { calculateWinner } from '../calculateWinner'
import { Themes } from '../ThemeContext'
import * as actions from './actions'

const initialState = {
    history: [{
        squares: Array(9).fill(null),
        xIsNext: true
    }],
    stepNumber: 0,
    isAsc: true,
    theme: Themes.light,
    allowModal: false
};

export default function reducer(state = initialState, action) {
    var newState = { ...state };

    if (action.type === actions.CLICK_CELL) {
        const current = newState.history[newState.stepNumber];
        if (current.squares[action.cell] || calculateWinner(current.squares)[0]) {
            return newState;
        }

        const squares = current.squares.slice();
        squares[action.cell] = current.xIsNext ? 'X' : 'O';
        
        // need to provide a new array to keep a function pure
        newState.history = newState.history
            .slice(0, newState.stepNumber + 1)
            .concat({
                squares: squares,
                xIsNext: !current.xIsNext,
                cell: action.cell
            });
        newState.stepNumber++;
        newState.allowModal = true;
    }

    if (action.type === actions.CHANGE_THEME) {
        newState.theme = Themes.oppositeOf(newState.theme);
    }

    if (action.type === actions.CHANGE_LOG_DIR) {
        newState.isAsc = !newState.isAsc;
    }

    if (action.type === actions.MOVE_TO_STEP) {
        newState.stepNumber = action.step;
    }

    if (action.type === actions.SET_ALLOW_MODAL) {
        newState.allowModal = action.isAllowed;
    }

    return newState;
}