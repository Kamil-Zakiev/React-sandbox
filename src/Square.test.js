import React from 'react'
import renderer from 'react-test-renderer';
import { Square } from './Square'
import click from './effects/click'
import { ClickCell } from './store/actionCreators';
import { ThemeContext } from './ThemeContext';
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16';

jest.mock('./effects/click')
configure({ adapter: new Adapter() });

describe('<Square />', () => {
    function prepareComponent(inWinDirection = false) {
        const props = {
            index: 5,
            dispatch: jest.fn(),
            inWinDirection: inWinDirection,
            value: 'X'
        };

        const component = renderer.create(
            <Square
                {...props}
            />
        );

        return {
            props,
            component
        };
    }

    beforeEach(() => {
        click.mockClear();
    });

    it('should render correctly', () => {
        const { component } = prepareComponent();
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should dispatch a ClickCell action when is clicked', () => {
        const props = {
            index: 5,
            dispatch: jest.fn(),
            inWinDirection: false,
            value: 'X'
        };

        const wrapper = shallow(<Square {...props} />);
        wrapper.find('button').simulate('click');

        expect(props.dispatch.mock.calls.length).toBe(1);
        expect(props.dispatch.mock.calls[0][0]).toEqual(ClickCell(props.index));
    });

    it('should apply click effect when is clicked', () => {        
        const props = {
            index: 5,
            dispatch: jest.fn(),
            inWinDirection: false,
            value: 'X'
        };

        const wrapper = shallow(<Square {...props} />);
        wrapper.find('button').simulate('click');
        
        expect(click.mock.calls.length).toBe(1);
    });

    it('should add theme based class', () => {
        const component = renderer.create(
            <ThemeContext.Provider value='test-theme'>
                <Square />
            </ThemeContext.Provider>
        );

        const json = component.toJSON();
        expect(json.props.className).toContain('test-theme-button');
    });

    it('Should add `winCell` when square is in win direction', () => {
        const { component } = prepareComponent(true);

        const json = component.toJSON();
        expect(json.props.className).toContain('winCell');
    });

    it('Should not add `winCell` when square is not in win direction', () => {
        const { component } = prepareComponent(false);

        const json = component.toJSON();
        expect(json.props.className).not.toContain('winCell');
    });

    it('Should have a `square` class', () => {
        const { component } = prepareComponent(false);

        const json = component.toJSON();
        expect(json.props.className).toContain('square');
    });
})