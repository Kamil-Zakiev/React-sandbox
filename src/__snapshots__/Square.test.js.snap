// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Square /> should render correctly 1`] = `
<button
  className="square false dark-button"
  onClick={[Function]}
>
  X
</button>
`;
